### ISSUE TITLE
Eclipselink entity listeners stored in java.util.Vector results in high thread contention

### ISSUE THREAD

**Issue by svendiedrichsen** [payara/Payara#1826]
# Description #
----------

Eclipselink bug https://bugs.eclipse.org/bugs/show_bug.cgi?id=291546

I'm sorry to be bothering you but I know you are using a patched version of Eclipselink for the Payara server and might want to fix this.

Trying to make use of entity listeners revealed that Eclipselink uses `java.util.Vector` and `java.util.Hashtable` to manage listeners which under higher system load results in high thread contention. In our case the system cpu load as well as the number of interrupts tripled.

Here is a typical stacktrace:

```
java.lang.Thread.State: BLOCKED
at java.util.Vector.size(Vector.java:315)
    waiting to lock <2ed58551> (a java.util.Vector) owned by "jk-main-thread-pool-1::jk-main-listener-1(41)" t@8675
    at org.eclipse.persistence.descriptors.DescriptorEventManager.notifyEJB30Listeners(DescriptorEventManager.java:672)
    at org.eclipse.persistence.descriptors.DescriptorEventManager.executeEvent(DescriptorEventManager.java:229)
    at org.eclipse.persistence.descriptors.changetracking.DeferredChangeDetectionPolicy.calculateChanges(DeferredChangeDetectionPolicy.java:87)
    at org.eclipse.persistence.descriptors.changetracking.DeferredChangeDetectionPolicy.calculateChangesForExistingObject(DeferredChangeDetectionPolicy.java:56)
    at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.calculateChanges(UnitOfWorkImpl.java:664)
    at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.commitToDatabaseWithChangeSet(UnitOfWorkImpl.java:1516)
    at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.issueSQLbeforeCompletion(UnitOfWorkImpl.java:3168)
    at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.issueSQLbeforeCompletion(RepeatableWriteUnitOfWork.java:355)
    at org.eclipse.persistence.transaction.AbstractSynchronizationListener.beforeCompletion(AbstractSynchronizationListener.java:158)
    at org.eclipse.persistence.transaction.JTASynchronizationListener.beforeCompletion(JTASynchronizationListener.java:68)
    at com.sun.enterprise.transaction.JavaEETransactionImpl.commit(JavaEETransactionImpl.java:452)
    at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.commit(JavaEETransactionManagerSimplified.java:879)
    at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:721)
    at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:505)
    at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4542)
    at com.sun.ejb.containers.StatefulSessionContainer.postInvokeTx(StatefulSessionContainer.java:1853)
    at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2021)
    at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:1992)
    at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)

```
## Expected Outcome

Listener notification does not extensively synchronize concurrent reader threads. Maybe use some un-synchronized list to notify listeners.

## Current Outcome

Listener notification has high potential for thread contention.

## Steps to reproduce (Only for bug reports) 

Have an App to register an entity listener and a lot of threads simultaniously fire notifications. 

## Environment ##

- **Payara Version**: 4.1.2.172
- **Edition**: Full
- **JDK Version**: 8 u91 - Oracle
- **Operating System**: Linux
- **Database**: Oracle

**Comment by lprimak** [comment]
I would open up a bug with EclipseLink as well

**Comment by fturizo** [comment]
@svendiedrichsen, do you have a sample reproducer for this Entity Listener scenario? Also as @lprimak recommended, try to report this issue on the EclipseLink project as well to get their input.

**Comment by svendiedrichsen** [comment]
There already is a bug filed with EclipseLink https://bugs.eclipse.org/bugs/show_bug.cgi?id=291546

It was filed on October 2010. So I don't think they will pay any attention anymore.

I will put together a test case and post it here.

**Comment by svendiedrichsen** [comment]
Here is another piece of code which is within Eclipselink class 'org.eclipse.persistence.internal.jpa.metadata.listeners.EntityListener' which cause massive thread contention:

'jk-main-thread-pool-1::jk-main-listener-1(2)" #52 daemon prio=5 os_prio=0 tid=0x00007f4908338000 nid=0x5458 waiting for monitor entry [0x00007f48f8ca0000]
java.lang.Thread.State: BLOCKED (on object monitor)
at java.util.Hashtable.containsKey(Hashtable.java:334)
waiting to lock <0x000000073f47c1c8> (a java.util.Hashtable)
at org.eclipse.persistence.internal.jpa.metadata.listeners.EntityListener.isOverriddenEvent(EntityListener.java:371)
at org.eclipse.persistence.descriptors.DescriptorEventManager.notifyEJB30Listeners(DescriptorEventManager.java:683)
at org.eclipse.persistence.descriptors.DescriptorEventManager.executeEvent(DescriptorEventManager.java:229)
at org.eclipse.persistence.descriptors.changetracking.DeferredChangeDetectionPolicy.calculateChanges(DeferredChangeDetectionPolicy.java:82)
at org.eclipse.persistence.descriptors.changetracking.DeferredChangeDetectionPolicy.calculateChangesForExistingObject(DeferredChangeDetectionPolicy.java:56)
at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.calculateChanges(UnitOfWorkImpl.java:664)
at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.commitToDatabaseWithChangeSet(UnitOfWorkImpl.java:1516)
at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.issueSQLbeforeCompletion(UnitOfWorkImpl.java:3168)
at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.issueSQLbeforeCompletion(RepeatableWriteUnitOfWork.java:355)
at org.eclipse.persistence.transaction.AbstractSynchronizationListener.beforeCompletion(AbstractSynchronizationListener.java:158)
at org.eclipse.persistence.transaction.JTASynchronizationListener.beforeCompletion(JTASynchronizationListener.java:68)
at com.sun.enterprise.transaction.JavaEETransactionImpl.commit(JavaEETransactionImpl.java:452)
at com.sun.enterprise.transaction.JavaEETransactionManagerSimplified.commit(JavaEETransactionManagerSimplified.java:879)
at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:721)
at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:505)
at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4542)
at com.sun.ejb.containers.StatefulSessionContainer.postInvokeTx(StatefulSessionContainer.java:1853)
at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2021)
at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:1992)
at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)'

**Comment by fturizo** [comment]
@svendiedrichsen, we have done fixes for EclipseLink's bugs before, so in this case, a fix from our development team is possible. However, due to the amount of effort needed to work for these fixes, we prioritize doing them for customers of paid support. If solving this issue is a priority for you, I recommend you to consider acquiring a support contract so we can proceed with haste on solving this issue for you. 

Other than that I could report this issue to our developers, but I can't guarantee whether or not will be fixed or if so, that it will be solved on a short-term basis.

**Comment by ggam** [comment]
Another chance would be to report this issue on the GlassFish repository, since they are EclipseLink commiters and in charge of the JPA RI.

**Comment by svendiedrichsen** [comment]
Added Glassfish bug https://github.com/javaee/glassfish/issues/22183 as suggested.

**Comment by svendiedrichsen** [comment]
If this is a question of resources than I'm willing to help out if I may. Is there a possibility for me to work on this?

**Comment by smillidge** [comment]
Absolutely we'd love you to work on it. If you can make the change in a branch of EclipseLink and test it out at high concurrency we'd be willing to pull the changes into a patched eclipselink version in Payara and work with you to get these back into upstream Eclipselink.

My main concern is the testing of the change so if you have the high concurrency environment set up that would be a great start.

**Comment by svendiedrichsen** [comment]
Great. I'll come back when I have the test environment working. May take some days time though as I'll do it in my spare time.

**Comment by arjantijms** [comment]
In general this type of change, replacing `Hashtable` and `Vector` among others, is something that's indeed important.

The difficult thing though is that we do need to prove that the replacement works and that the synchronisation was either not needed at all, or that a new synchronisation mechanism doesn't introduce new bugs or races.

**Comment by svendiedrichsen** [comment]
Depending on the importance of performance for your system you may find the replacement nessessary or not. Our system is predicated on performance so we judge the replacement to be essential.

Software development (at least as it is done generally) is not math so we don't have to prove anything. We only need to have the tests working.

**Comment by svendiedrichsen** [comment]
# I have made the eclipselink changes in branch 2.6 under https://github.com/svendiedrichsen/eclipselink.runtime
# Under https://github.com/svendiedrichsen/eclipselink-entity-listener you will find test scenarios. Especially test-with-both-listeners has several entities with both types of listeners. Internal ones within the entities and externally registered with the container.
# Under gatling-test you will find a test scenario which targets the test-with-both-listeners web application and runs for 30 minutes and produces some load.

You need to have a server with several cores to run the server with the test app. The more cores the more you will find blocked threads. To achive best visibility of this contention you need to configure the payara server to
# have as many {{acceptor-threads}} as you have cores
# use a non-synched thread-pool thus configuring the thread-pool to have {{min-threads=-1}} and a {{queue-size=0}}

Hope this may help.

**Comment by svendiedrichsen** [comment]
It looks like my PR https://github.com/eclipse/eclipselink.runtime/pull/30 for this bug is recognized and will be applied.

**Comment by svendiedrichsen** [comment]
According to @kratz74 the fix is in the trunk and will be most likely also be ported to 2.7 branch. I was kindly asking him to have it also backported to 2.6.

**Comment by svendiedrichsen** [comment]
The changes are accepted and pulled:
> Fix is now in master, 2.7 and 2.6 branches. It will be available when 2.6.6 is released.
See https://bugs.eclipse.org/bugs/show_bug.cgi?id=291546 for more details.
We'll also integrate it into GlassFish4.

**Comment by svendiedrichsen** [comment]
If I may I would like to kindly ask you to pull the Eclipselink 2.6.6 changes into Payara 4 when they are relreased and available.
